packages/liveblocks-core/src/lib/AsyncCache.ts

```diff
@@ -0,0 +1,345 @@
+import type { Callback, Observable, UnsubscribeCallback } from "./EventSource";
+import { makeEventSource } from "./EventSource";
+
+const DEDUPLICATION_INTERVAL = 2000;
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };
+
+type AsyncFunction<T, A extends any[] = any[]> = (...args: A) => Promise<T>;
+
+type AsyncCacheOptions = {
+  deduplicationInterval?: number;
+};
+
+type AsyncCacheItemOptions = WithRequired<
+  AsyncCacheOptions,
+  "deduplicationInterval"
+>;
+
+type InvalidateOptions<TData = any> =
+  | { clearData?: false; setOptimisticData?: never }
+  | {
+      clearData?: never;
+      setOptimisticData: (data: TData | undefined) => TData | undefined;
+    };
+
+export type AsyncState<TData = any, TError = any> = {
```

Comment:
Can we model this as a three-legged ADT?

```
type AsyncState<T, E> =
  | AsyncStateLoading
  | AsyncStateResolved<T>
  | AsyncStateErorr<E>

type AsyncStateLoading = {
  readonly loading: true;
  readonly data?: never;
  readonly error?: never;
};

type AsyncStateResolve<T> = {
  readonly loading: false;
  readonly data: T;
  readonly error?: never;
};

type AsyncStateError<E> = {
  readonly loading: false;
  readonly data?: never;
  readonly error?: E;
};
```

This will make any impossible state impossible to express at the type level. The object `{ isLoading: false, data: undefined, error: undefined }` or `{ isLoading: true, data, error }` are currently possible, but should be invalid, right?

Reply:
Also, I would advise to never use default type arguments in TypeScript, especially when defaulting to `any`. It makes it way too easy to end up with unintended `any` types values. Why not remove them? Anywhere you use the `AsyncState` type, you should specify what its expected "ok" and "error" type payloads are. Like any other container, think a Map or Array, it would also be weird to use them without their payload types, right?

Reply:
Lastly, when writing _generic_ data structures, I wouldn't mind using single-letter type variables like `T` and `E`. Just like how `Array<T>` and `Map<K, V>` are used in the TypeScript standard library. They may make the rest of the code in this module more concice.

Reply:
I've refactored things to use a discriminated union but because of the API's genericness (to support vanilla and React use-cases) and options like `clearData` and `optimisticData`, there are 4 possible states with some overlap: `AsyncStateInvalid`, `AsyncStateLoading`, `AsyncStateSuccess`, and `AsyncStateError`.